presto-main/src/main/java/io/prestosql/sql/analyzer/Analysis.java

```diff
@@ -646,21 +649,37 @@ public JoinUsingAnalysis getJoinUsing(Join node)
 
     public void addTableColumnReferences(AccessControl accessControl, Identity identity, Multimap<QualifiedObjectName, String> tableColumnMap)
     {
-        AccessControlInfo accessControlInfo = new AccessControlInfo(accessControl, identity);
-        Map<QualifiedObjectName, Set<String>> references = tableColumnReferences.computeIfAbsent(accessControlInfo, k -> new LinkedHashMap<>());
-        tableColumnMap.asMap()
-                .forEach((key, value) -> references.computeIfAbsent(key, k -> new HashSet<>()).addAll(value));
+        for (QualifiedObjectName objectName : tableColumnMap.keySet()) {
+            addTableColumnAccess(accessControl, identity, objectName, ImmutableSet.copyOf(tableColumnMap.get(objectName)));
+        }
     }
 
     public void addEmptyColumnReferencesForTable(AccessControl accessControl, Identity identity, QualifiedObjectName table)
     {
-        AccessControlInfo accessControlInfo = new AccessControlInfo(accessControl, identity);
-        tableColumnReferences.computeIfAbsent(accessControlInfo, k -> new LinkedHashMap<>()).computeIfAbsent(table, k -> new HashSet<>());
+        addTableColumnAccess(accessControl, identity, table, ImmutableSet.of());
+    }
+
+    private void addTableColumnAccess(AccessControl accessControl, Identity identity, QualifiedObjectName objectName, Set<String> columns)
+    {
+        if (tableColumnAccesses.containsKey(objectName)) {
+            Optional<AccessControlCheck> oldTableAccessControlCheck = tableColumnAccesses.get(objectName).stream()
+                    .filter(accessControlCheck -> accessControlCheck.isUsing(accessControl, identity))
```

Comment:
You manually created a map from
`(objectName (eq), accessControl (==), identity (==))` to `columns`

why not define such a key class explicitly and use regular map?

Reply:
also, this means that if a have a query using multiple VIEWs all based on same table, table/column checks will probably be invoked multiple times, since each time a VIEW is accessed, a new `Identity` (for view definer) will be instantiated. 

Can you verify whether this is really the case? This could cause access control checks amplification...
Can we have a test ensuring this is not the case?